### PR TITLE
Direct 1:1 port pandemic.dm from Hatestation. Credits to Koriath for coding it.

### PR DESCRIPTION
**Pull request by LanCartwright**
:cl: PANDEMIC Port.
add: Inert virus
tweak: Changed PANDEMIC to ported from hatestation.
add: Ability to revert your virus to an inert virus on the machine.
add: Ability to add plasma to the machine or to the virus through the machine.
add: Ability to add mutagen to the machine or to the virus through the machine.
add: Ability to add synaptizine to the machine or to the virus through the machine.
add: Ability to add virus food to the machine or to the virus through the machine.
add: Ability to add and load symptoms. from a database of already discovered symptoms.
/:cl:

Credits to @Koriath for this.

### CODE REVIEW

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 1]
```diff
@@ -52,15 +64,15 @@
...
 	if(stat & BROKEN)
-		icon_state = (beaker ? "mixer1_b" : "mixer0_b")
+		icon_state = (src.beaker?"mixer1_b":"mixer0_b")
 		return
 
 	icon_state = "mixer[(beaker)?"1":"0"][(powered()) ? "" : "_nopower"]"
 
 	if(wait)
-		cut_overlays()
+		overlays.Cut()
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: you have to use our specific methods for adding and removing overlays

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 2]
```diff
@@ -167,124 +273,228 @@
 	src.add_fingerprint(usr)
 	return
 
-/obj/machinery/computer/pandemic/attack_hand(mob/user)
+/obj/machinery/computer/pandemic/attack_hand(mob/user as mob)
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: no

**Comment by optimumtact** [comment]
You didn't port this at all, this is just a copy paste. Don't waste our time with this nonsense

**Comment by LanCartwright** [comment]
Whether it's a copypaste or not, it's better than what is currently avaliable and will be better than what Coffescript will do.
This is a direct 1:1 port with errors smoothed out.

Please raise concerns on what isn't right or wrong and I'll change them. It's appreciated to note that this wasn't my creation. I was simply the host.

**Comment by optimumtact** [comment]
You can't just copy and paste code from another server, you have to adjust all the specifics to match with our codebase and code standards. As maintainers we shouldn't have to go through and point them all out to you first, most of them should already be spotted and fixed before we review a pr.

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 3]
```diff
@@ -52,15 +64,15 @@
 
 /obj/machinery/computer/pandemic/update_icon()
 	if(stat & BROKEN)
-		icon_state = (beaker ? "mixer1_b" : "mixer0_b")
+		icon_state = (src.beaker?"mixer1_b":"mixer0_b")
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: please restore the spacing and remove the unnecessary src

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 4]
```diff
@@ -36,13 +48,13 @@
...
 
-/obj/machinery/computer/pandemic/proc/GetVirusTypeByIndex(index)
+/obj/machinery/computer/pandemic/proc/GetVirusTypeByIndex(var/index)
 	var/datum/disease/D = GetVirusByIndex(index)
 	if(D)
 		return D.GetDiseaseID()
 	return null
 
-/obj/machinery/computer/pandemic/proc/replicator_cooldown(waittime)
+/obj/machinery/computer/pandemic/proc/replicator_cooldown(var/waittime)
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: we don't usually put var/ in our method declarations

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 5]
```diff
@@ -69,48 +81,45 @@
...
+	if (href_list["symptom"])
+		if(beaker && beaker.reagents)
+			if(beaker.reagents.reagent_list.len)
+				var/datum/reagent/blood/BL = locate() in beaker.reagents.reagent_list
+				if(BL)
+					if(BL.data && BL.data["viruses"])
+						var/list/viruses = BL.data["viruses"]
+						for(var/datum/disease/advance/D in viruses)
+							D.AddSymptom(new_symptoms[text2num(href_list["symptom"])])
+		src.updateUsrDialog()
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: remove the unnecessary src in these calls as well (same goes for further examples below)

**Comment by LanCartwright** [comment]
Alright. My apologies. I'll ask for help in future.

Nobody else was gonna do this. And I knew I could get help.

I mean if you don't want it, I'll drop it and leave it at that. But this is code going to waste from an otherwise less "sophisticated" codebase with more optimistic coders trying things otherwise wouldn't be tried.

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 6]
```diff
@@ -158,7 +154,117 @@
...
+								if(D.IsSame(DT))
+									d_test = 0
+							if(d_test)
+								new_diseases += D
+								usr << "New disease added to the database!"
+	else if(href_list["update_symptom"])
+		if(beaker && beaker.reagents)
+			if(beaker.reagents.reagent_list.len)
+				var/datum/reagent/blood/BL = locate() in beaker.reagents.reagent_list
+				if(BL)
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: this block of four checks is repeated multiple times in this topic call

split it out into it's own proc that returns true if all four checks pass or false if they don’t


    is_valid_beaker  

and call it instead of retyping these four checks over and over

    else if(href_list["update_symptom"])
        if(is_valid_beaker(beaker))
            do thing

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 7]
```diff
@@ -167,124 +273,228 @@
...
 
-		beaker =  I
-		beaker.loc = src
-		user << "<span class='notice'>You add the beaker to the machine.</span>"
+		if(src.beaker)
+			user << "A beaker is already loaded into the machine."
+			return
+		src.beaker =  I
+		user.drop_item()
+		I.loc = src
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: force move instead of just updating the loc

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 8]
```diff
@@ -167,124 +273,228 @@
...
+		user << "You add the beaker to the machine!"
 		src.updateUsrDialog()
 		icon_state = "mixer1"
+
+	else if(istype(I, /obj/item/weapon/screwdriver))
+		if(src.beaker)
+			beaker.loc = get_turf(src)
```
Comment (optimumtact) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: again use forceMove proc

**Comment by optimumtact** [comment]
It's not that I don't want it, frankly I'm ambivalent on it as a change, it won't have any effect on me personally in game.

I just want to see a little more care and effort applied when porting things is all.

**Comment by phil235** [comment]
This isn't a port, it's a copypaste...

**Comment by AnturK** [comment]
Dynamic data single screen interface. Not TGUI. Please.

File: code/modules/reagents/chemistry/machinery/pandemic.dm (hunk 4 above)
Comment (RemieRichards) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: we actively work against it, actually.

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 9]
```diff
@@ -52,15 +64,15 @@
...
+		icon_state = (src.beaker?"mixer1_b":"mixer0_b")
 		return
 
 	icon_state = "mixer[(beaker)?"1":"0"][(powered()) ? "" : "_nopower"]"
 
 	if(wait)
 		cut_overlays()
 	else
-		add_overlay("waitlight")
+		overlays += "waitlight"
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: revert this change.

File: code/modules/reagents/chemistry/machinery/pandemic.dm [hunk 10]
```diff
@@ -69,48 +81,45 @@
 	usr.set_machine(src)
 	if(!beaker) return
 
-	if (href_list["create_vaccine"])
+	if (href_list["symptom"])
+		if(beaker && beaker.reagents)
+			if(beaker.reagents.reagent_list.len)
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/machinery/pandemic.dm]: this and the above can be one line.

**Comment by ike709** [comment]
As someone who actually plays virologist sometimes, allowing people to add plasma/mutagen/etc. through the machine would be an _amazing_ quality of life change. :+1: x :100: 

The actual code quality of this PR though, is another story.

**Comment by Koriath** [comment]
I wrote this a long time ago, before tgui was a thing.
It still has a bunch of problems, and the Interface could be organized better (especially the symptom database).
Do note however that it had two main issues I never fixed:

>Creating Vaccines is impossible
I mean, you can do it the old fashioned way, but even after storing it in the machine, it fails to replicate it and just prints an empty bottle. I personally blame the way diseases/blood data is stored.

>It is, without a better word, overpowered
Main problem might be the base Virology gameplay though. This pandemic version just accelerated it.
It became a 5 minute effort to make 6 symptom Virus (either beneficial or harmfull) especially with the possibility of saving symptoms in a database.
If that's something you guys might like, go for it. Test it out. But I'm predicting a lot of epidemics for a few days, so someone might wanna balance this out a little better in the future.

**Comment by rouge** [comment]
>merging hatestation stuff into upstream

**Comment by KorPhaeron** [comment]
@LanCartwright 

Week without changes and doesn't compile, let me know if you're going to finish this and I'll reopen
